cornac/eval_methods/stratified_split.py

```diff
@@ -0,0 +1,122 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ============================================================================
+
+from math import ceil
+
+from .base_method import BaseMethod
+from ..utils import get_rng
+from ..utils.common import safe_indexing
+
+
+class StratifiedSplit(BaseMethod):
+    """Grouping data by user or item then splitting data into training, validation, and test sets.
+
+    Parameters
+    ----------
+    data: array-like, required
+        Raw preference data in the triplet format [(user_id, item_id, rating_value, review_time)].
+
+    group_by: str, optional, default: 'user'
+        Group option: either 'user' or 'item'
+
+    chrono: bool, optional, default False
+        Data is ordered by reviewed time or not. If this option is True, data must be in 'UIRT' format.
+
+    test_size: float, optional, default: 0.2
+        The proportion of the test set.
+
+    val_size: float, optional, default: 0.0
+        The proportion of the validation set.
+
+    rating_threshold: float, optional, default: 1.0
+        Threshold used to binarize rating values into positive or negative feedback for
+        model evaluation using ranking metrics (rating metrics are not affected).
+
+    seed: int, optional, default: None
+        Random seed for reproducibility.
+
+    exclude_unknowns: bool, optional, default: True
+        If `True`, unknown users and items will be ignored during model evaluation.
+
+    verbose: bool, optional, default: False
+        Output running log.
+
+    """
+
+    def __init__(self, data, group_by='user', chrono=False, fmt='UIRT', test_size=0.2, val_size=0.0, rating_threshold=1.0,
+                 seed=None, exclude_unknowns=True, verbose=False, **kwargs):
+        super().__init__(data=data, fmt=fmt, rating_threshold=rating_threshold, seed=seed,
+                         exclude_unknowns=exclude_unknowns, verbose=verbose, **kwargs)
+        if group_by not in ['user', 'item']:
+            raise ValueError("group_by option must be either 'user' or 'item' but {}".format(group_by))
+        if chrono and (fmt != 'UIRT' or len(self._data[0]) != 4):
+            raise ValueError('Input data must be in "UIRT" format')
+        self.chrono = chrono
+        self.group_by = group_by
+        self.train_size, self.val_size, self.test_size = self.validate_size(val_size, test_size)
+        self._split()
+
+    @staticmethod
+    def validate_size(val_size, test_size):
+        if val_size is None:
+            val_size = 0.0
+        elif val_size < 0 or val_size > 1.0:
```

Comment:
Can we also accept `val_size` and `test_size` of integer values? That would be more practical.

Reply:
validate_size is called before split for checking the input ratio. Each group of user/item may have different number of ratings, so it would be difficult to use integer in this case.

Reply:
in this case, `validate_size` will be called each time we split data for a user/item